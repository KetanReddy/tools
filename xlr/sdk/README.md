# @player-tools/xlr-sdk

Unified interface for working with XLRs. Abstracts away needing to know how they are packaged or where they are sourced from. Current operations supported: 

- Load
- Export
  - Typescript
- Validate

Also exposes functionality to create your own XLR Registry to manage how the SDK stores loaded types. 