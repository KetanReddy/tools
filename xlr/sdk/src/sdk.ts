import type {
  Manifest,
  NamedType,
  NodeType,
  TransformFunction,
  TSManifest,
} from '@player-tools/xlr';
import type { TopLevelDeclaration } from '@player-tools/xlr-utils';
import { fillInGenerics } from '@player-tools/xlr-utils';
import type { Node } from 'jsonc-parser';
import { TSWriter } from '@player-tools/xlr-converters';
import fs from 'fs';
import path from 'path';
import ts from 'typescript';

import type { XLRRegistry, Filters } from './registry';
import { BasicXLRRegistry } from './registry';
import type { ExportTypes } from './types';
import { XLRValidator } from './validator';

/**
 * Abstraction for interfacing with XLRs making it more approachable to use without understanding the inner workings of the types and how they are packaged
 */
export class XLRSDK {
  private registry: XLRRegistry;
  private validator: XLRValidator;
  private tsWriter: TSWriter;

  constructor(customRegistry?: XLRRegistry) {
    this.registry = customRegistry ?? new BasicXLRRegistry();
    this.validator = new XLRValidator(this.registry);
    this.tsWriter = new TSWriter();
  }

  public loadDefinitionsFromDisk(
    inputPath: string,
    filters?: Omit<Filters, 'pluginFilter'>,
    transforms?: Array<TransformFunction>
  ) {
    const manifest = JSON.parse(
      fs.readFileSync(path.join(inputPath, 'xlr', 'manifest.json')).toString(),
      (key: unknown, value: unknown) => {
        // Custom parser because JSON objects -> JS Objects, not maps
        if (typeof value === 'object' && value !== null) {
          if (key === 'capabilities') {
            return new Map(Object.entries(value));
          }
        }

        return value;
      }
    ) as Manifest;

    manifest.capabilities?.forEach((capabilityList, capabilityName) => {
      if (
        filters?.capabilityFilter &&
        capabilityName.match(filters?.capabilityFilter)
      )
        return;
      capabilityList.forEach((extensionName) => {
        if (!filters?.typeFilter || !extensionName.match(filters?.typeFilter)) {
          const cType: NamedType<NodeType> = JSON.parse(
            fs
              .readFileSync(
                path.join(inputPath, 'xlr', `${extensionName}.json`)
              )
              .toString()
          );
          transforms?.forEach((transform) => transform(cType, capabilityName));
          const resolvedType = fillInGenerics(cType) as NamedType<NodeType>;
          this.registry.add(resolvedType, manifest.pluginName, capabilityName);
        }
      });
    });
  }

  public async loadDefinitionsFromModule(
    inputPath: string,
    filters?: Omit<Filters, 'pluginFilter'>,
    transforms?: Array<TransformFunction>
  ) {
    const importManifest = await import(
      path.join(inputPath, 'xlr', 'manifest.js')
    );
    const manifest = importManifest.default as TSManifest;

    Object.keys(manifest.capabilities)?.forEach((capabilityName) => {
      if (
        filters?.capabilityFilter &&
        capabilityName.match(filters?.capabilityFilter)
      )
        return;
      const capabilityList = manifest.capabilities[capabilityName];
      capabilityList.forEach((extension) => {
        if (
          !filters?.typeFilter ||
          !extension.name.match(filters?.typeFilter)
        ) {
          transforms?.forEach((transform) =>
            transform(extension, extension.name)
          );
          const resolvedType = fillInGenerics(extension) as NamedType<NodeType>;
          this.registry.add(resolvedType, manifest.pluginName, extension.name);
        }
      });
    });
  }

  public getType(id: string) {
    return this.registry.get(id);
  }

  public hasType(id: string) {
    return this.registry.has(id);
  }

  public listTypes(filters?: Filters) {
    return this.registry.list(filters);
  }

  public validate(typeName: string, rootNode: Node) {
    const xlr = this.registry.get(typeName);
    if (!xlr) {
      throw new Error(
        `Type ${typeName} does not exist in registry, can't validate`
      );
    }

    return this.validator.validateType(rootNode, xlr);
  }

  /**
   * Exports the types loaded into the registry to the specified format
   *
   * @param exportType - what format to export as
   * @param importMap - a map of primitive packages to types exported from that package to add import statements
   * @param filters - filter out plugins/capabilities/types you don't want to export
   * @param transforms - transforms to apply to types before exporting them
   * @returns [filename, content][] - Tuples of filenames and content to write
   */
  public exportRegistry(
    exportType: ExportTypes,
    importMap: Map<string, string[]>,
    filters?: Filters,
    transforms?: Array<TransformFunction>
  ): [string, string][] {
    const typesToExport = this.registry.list(filters).map((type) => {
      transforms?.forEach((transformFunction) =>
        transformFunction(
          type,
          this.registry.info(type.name)?.capability as string
        )
      );
      return type;
    });

    if (exportType === 'TypeScript') {
      const outputString = this.exportToTypeScript(typesToExport, importMap);
      return [['out.d.ts', outputString]];
    }

    throw new Error(`Unknown export format ${exportType}`);
  }

  private exportToTypeScript(
    typesToExport: NamedType[],
    importMap: Map<string, string[]>
  ): string {
    const referencedImports: Set<string> = new Set();
    const exportedTypes: Map<string, TopLevelDeclaration> = new Map();
    const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });

    let resultFile = ts.createSourceFile(
      'output.d.ts',
      '',
      ts.ScriptTarget.ES2017,
      false, // setParentNodes
      ts.ScriptKind.TS
    );

    typesToExport.forEach((typeNode) => {
      const { type, referencedTypes, additionalTypes } =
        this.tsWriter.convertNamedType(typeNode);
      exportedTypes.set(typeNode.name, type);
      additionalTypes?.forEach((additionalType, name) =>
        exportedTypes.set(name, additionalType)
      );
      referencedTypes?.forEach((referencedType) =>
        referencedImports.add(referencedType)
      );
    });

    const typesToPrint: Array<string> = [];

    exportedTypes.forEach((type) =>
      typesToPrint.push(
        printer.printNode(ts.EmitHint.Unspecified, type, resultFile)
      )
    );

    importMap.forEach((imports, packageName) => {
      const applicableImports = imports.filter((i) => referencedImports.has(i));
      resultFile = ts.factory.updateSourceFile(resultFile, [
        ts.factory.createImportDeclaration(
          /* decorators */ undefined,
          /* modifiers */ undefined,
          ts.factory.createImportClause(
            false,
            undefined,
            ts.factory.createNamedImports(
              applicableImports.map((i) =>
                ts.factory.createImportSpecifier(
                  undefined,
                  ts.factory.createIdentifier(i)
                )
              )
            )
          ),
          ts.factory.createStringLiteral(packageName)
        ),
        ...resultFile.statements,
      ]);
    });

    const headerText = printer.printFile(resultFile);
    const nodeText = typesToPrint.join('\n');
    return `${headerText}\n${nodeText}`;
  }
}
