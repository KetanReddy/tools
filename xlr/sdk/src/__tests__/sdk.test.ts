import type { TransformFunction } from '@player-tools/xlr';
import { parseTree } from 'jsonc-parser';
import type { Filters } from '../registry';
import { XLRSDK } from '../sdk';

const EXCLUDE: Filters = { typeFilter: 'Transformed' };

describe('Loading XLRs', () => {
  it('Loading from Disk', () => {
    const sdk = new XLRSDK();
    sdk.loadDefinitionsFromDisk('./common/static_xlrs/plugin', EXCLUDE);
    sdk.loadDefinitionsFromDisk('./common/static_xlrs/core');

    expect(sdk.hasType('Asset')).toStrictEqual(true);
    expect(sdk.hasType('AssetWrapper')).toStrictEqual(true);
    expect(sdk.hasType('Binding')).toStrictEqual(true);
    expect(sdk.hasType('BindingRef')).toStrictEqual(true);
    expect(sdk.hasType('ExpressionRef')).toStrictEqual(true);
    expect(sdk.hasType('ActionAsset')).toStrictEqual(true);
    expect(sdk.hasType('InputAsset')).toStrictEqual(true);
    expect(sdk.hasType('TransformedAction')).toStrictEqual(false);
    expect(sdk.hasType('TransformedInput')).toStrictEqual(false);
  });

  it('Loading from Module', async () => {
    const sdk = new XLRSDK();
    await sdk.loadDefinitionsFromModule(
      '../../../common/static_xlrs/plugin',
      EXCLUDE
    );
    await sdk.loadDefinitionsFromModule('../../../common/static_xlrs/core');

    expect(sdk.hasType('Asset')).toStrictEqual(true);
    expect(sdk.hasType('AssetWrapper')).toStrictEqual(true);
    expect(sdk.hasType('Binding')).toStrictEqual(true);
    expect(sdk.hasType('BindingRef')).toStrictEqual(true);
    expect(sdk.hasType('ExpressionRef')).toStrictEqual(true);
    expect(sdk.hasType('ActionAsset')).toStrictEqual(true);
    expect(sdk.hasType('InputAsset')).toStrictEqual(true);
    expect(sdk.hasType('TransformedAction')).toStrictEqual(false);
    expect(sdk.hasType('TransformedInput')).toStrictEqual(false);
  });
});

describe('Basic Validation', () => {
  it('Working Test', () => {
    const mockAsset = parseTree(`
    {
      "id": 1,
      "type": "input",
      "binding": "some.data",
      "label": {
        "asset": {
          "value": "{{input.label}}"
        }
      }
    `);

    const sdk = new XLRSDK();
    sdk.loadDefinitionsFromDisk('./common/static_xlrs/plugin', EXCLUDE);
    sdk.loadDefinitionsFromDisk('./common/static_xlrs/core');

    expect(sdk.validate('InputAsset', mockAsset)).toMatchSnapshot();
  });
});

describe('Export Test', () => {
  it('Exports Typescript types', () => {
    const importMap = new Map([
      [
        '@player-ui/types',
        ['Expression', 'Asset', 'Binding', 'AssetWrapper', 'Schema.DataType'],
      ],
    ]);

    const sdk = new XLRSDK();
    sdk.loadDefinitionsFromDisk('./common/static_xlrs/plugin', EXCLUDE);
    const results = sdk.exportRegistry('TypeScript', importMap);
    expect(results[0][0]).toBe('out.d.ts');
    expect(results[0][1]).toMatchSnapshot();
  });

  it('Exports Typescript Types With Filters', () => {
    const importMap = new Map([
      [
        '@player-ui/types',
        ['Expression', 'Asset', 'Binding', 'AssetWrapper', 'Schema.DataType'],
      ],
    ]);

    const sdk = new XLRSDK();
    sdk.loadDefinitionsFromDisk('./common/static_xlrs/plugin', EXCLUDE);
    const results = sdk.exportRegistry('TypeScript', importMap, {
      typeFilter: 'Transformed',
    });
    expect(results[0][0]).toBe('out.d.ts');
    expect(results[0][1]).toMatchSnapshot();
  });

  it('Exports Typescript Types With Transforms', () => {
    const importMap = new Map([
      [
        '@player-ui/types',
        ['Expression', 'Asset', 'Binding', 'AssetWrapper', 'Schema.DataType'],
      ],
    ]);

    /**
     *
     */
    const transformFunction: TransformFunction = (input) => {
      if (input.type === 'object') {
        // eslint-disable-next-line no-param-reassign
        input.properties.transformed = {
          required: false,
          node: { type: 'boolean', const: true },
        };
      }
    };

    const sdk = new XLRSDK();
    sdk.loadDefinitionsFromDisk('./common/static_xlrs/plugin', EXCLUDE);
    const results = sdk.exportRegistry('TypeScript', importMap, {}, [
      transformFunction,
    ]);
    expect(results[0][0]).toBe('out.d.ts');
    expect(results[0][1]).toMatchSnapshot();
  });
});
