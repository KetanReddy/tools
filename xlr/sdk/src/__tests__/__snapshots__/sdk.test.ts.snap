// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Validation Working Test 1`] = `
Array [
  Object {
    "message": "Expected type 'string' but got 'number'",
    "node": Object {
      "children": Array [
        Object {
          "length": 4,
          "offset": 13,
          "parent": [Circular],
          "type": "string",
          "value": "id",
        },
        Object {
          "length": 1,
          "offset": 19,
          "parent": [Circular],
          "type": "number",
          "value": 1,
        },
      ],
      "colonOffset": 17,
      "length": 7,
      "offset": 13,
      "parent": Object {
        "children": Array [
          [Circular],
          Object {
            "children": Array [
              Object {
                "length": 6,
                "offset": 28,
                "parent": [Circular],
                "type": "string",
                "value": "type",
              },
              Object {
                "length": 7,
                "offset": 36,
                "parent": [Circular],
                "type": "string",
                "value": "input",
              },
            ],
            "colonOffset": 34,
            "length": 15,
            "offset": 28,
            "parent": [Circular],
            "type": "property",
          },
          Object {
            "children": Array [
              Object {
                "length": 9,
                "offset": 51,
                "parent": [Circular],
                "type": "string",
                "value": "binding",
              },
              Object {
                "length": 11,
                "offset": 62,
                "parent": [Circular],
                "type": "string",
                "value": "some.data",
              },
            ],
            "colonOffset": 60,
            "length": 22,
            "offset": 51,
            "parent": [Circular],
            "type": "property",
          },
          Object {
            "children": Array [
              Object {
                "length": 7,
                "offset": 81,
                "parent": [Circular],
                "type": "string",
                "value": "label",
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "length": 7,
                        "offset": 100,
                        "parent": [Circular],
                        "type": "string",
                        "value": "asset",
                      },
                      Object {
                        "children": Array [
                          Object {
                            "children": Array [
                              Object {
                                "length": 7,
                                "offset": 121,
                                "parent": [Circular],
                                "type": "string",
                                "value": "value",
                              },
                              Object {
                                "length": 17,
                                "offset": 130,
                                "parent": [Circular],
                                "type": "string",
                                "value": "{{input.label}}",
                              },
                            ],
                            "colonOffset": 128,
                            "length": 26,
                            "offset": 121,
                            "parent": [Circular],
                            "type": "property",
                          },
                        ],
                        "length": 48,
                        "offset": 109,
                        "parent": [Circular],
                        "type": "object",
                      },
                    ],
                    "colonOffset": 107,
                    "length": 57,
                    "offset": 100,
                    "parent": [Circular],
                    "type": "property",
                  },
                ],
                "length": 75,
                "offset": 90,
                "parent": [Circular],
                "type": "object",
              },
            ],
            "colonOffset": 88,
            "length": 84,
            "offset": 81,
            "parent": [Circular],
            "type": "property",
          },
        ],
        "length": 165,
        "offset": 5,
        "type": "object",
      },
      "type": "property",
    },
    "type": "type",
  },
  Object {
    "message": "Property 'id' missing from type 'Asset'",
    "node": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "length": 7,
              "offset": 121,
              "parent": [Circular],
              "type": "string",
              "value": "value",
            },
            Object {
              "length": 17,
              "offset": 130,
              "parent": [Circular],
              "type": "string",
              "value": "{{input.label}}",
            },
          ],
          "colonOffset": 128,
          "length": 26,
          "offset": 121,
          "parent": [Circular],
          "type": "property",
        },
      ],
      "length": 48,
      "offset": 109,
      "parent": Object {
        "children": Array [
          Object {
            "length": 7,
            "offset": 100,
            "parent": [Circular],
            "type": "string",
            "value": "asset",
          },
          [Circular],
        ],
        "colonOffset": 107,
        "length": 57,
        "offset": 100,
        "parent": Object {
          "children": Array [
            [Circular],
          ],
          "length": 75,
          "offset": 90,
          "parent": Object {
            "children": Array [
              Object {
                "length": 7,
                "offset": 81,
                "parent": [Circular],
                "type": "string",
                "value": "label",
              },
              [Circular],
            ],
            "colonOffset": 88,
            "length": 84,
            "offset": 81,
            "parent": Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "length": 4,
                      "offset": 13,
                      "parent": [Circular],
                      "type": "string",
                      "value": "id",
                    },
                    Object {
                      "length": 1,
                      "offset": 19,
                      "parent": [Circular],
                      "type": "number",
                      "value": 1,
                    },
                  ],
                  "colonOffset": 17,
                  "length": 7,
                  "offset": 13,
                  "parent": [Circular],
                  "type": "property",
                },
                Object {
                  "children": Array [
                    Object {
                      "length": 6,
                      "offset": 28,
                      "parent": [Circular],
                      "type": "string",
                      "value": "type",
                    },
                    Object {
                      "length": 7,
                      "offset": 36,
                      "parent": [Circular],
                      "type": "string",
                      "value": "input",
                    },
                  ],
                  "colonOffset": 34,
                  "length": 15,
                  "offset": 28,
                  "parent": [Circular],
                  "type": "property",
                },
                Object {
                  "children": Array [
                    Object {
                      "length": 9,
                      "offset": 51,
                      "parent": [Circular],
                      "type": "string",
                      "value": "binding",
                    },
                    Object {
                      "length": 11,
                      "offset": 62,
                      "parent": [Circular],
                      "type": "string",
                      "value": "some.data",
                    },
                  ],
                  "colonOffset": 60,
                  "length": 22,
                  "offset": 51,
                  "parent": [Circular],
                  "type": "property",
                },
                [Circular],
              ],
              "length": 165,
              "offset": 5,
              "type": "object",
            },
            "type": "property",
          },
          "type": "object",
        },
        "type": "property",
      },
      "type": "object",
    },
    "type": "missing",
  },
  Object {
    "message": "Property 'type' missing from type 'Asset'",
    "node": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "length": 7,
              "offset": 121,
              "parent": [Circular],
              "type": "string",
              "value": "value",
            },
            Object {
              "length": 17,
              "offset": 130,
              "parent": [Circular],
              "type": "string",
              "value": "{{input.label}}",
            },
          ],
          "colonOffset": 128,
          "length": 26,
          "offset": 121,
          "parent": [Circular],
          "type": "property",
        },
      ],
      "length": 48,
      "offset": 109,
      "parent": Object {
        "children": Array [
          Object {
            "length": 7,
            "offset": 100,
            "parent": [Circular],
            "type": "string",
            "value": "asset",
          },
          [Circular],
        ],
        "colonOffset": 107,
        "length": 57,
        "offset": 100,
        "parent": Object {
          "children": Array [
            [Circular],
          ],
          "length": 75,
          "offset": 90,
          "parent": Object {
            "children": Array [
              Object {
                "length": 7,
                "offset": 81,
                "parent": [Circular],
                "type": "string",
                "value": "label",
              },
              [Circular],
            ],
            "colonOffset": 88,
            "length": 84,
            "offset": 81,
            "parent": Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "length": 4,
                      "offset": 13,
                      "parent": [Circular],
                      "type": "string",
                      "value": "id",
                    },
                    Object {
                      "length": 1,
                      "offset": 19,
                      "parent": [Circular],
                      "type": "number",
                      "value": 1,
                    },
                  ],
                  "colonOffset": 17,
                  "length": 7,
                  "offset": 13,
                  "parent": [Circular],
                  "type": "property",
                },
                Object {
                  "children": Array [
                    Object {
                      "length": 6,
                      "offset": 28,
                      "parent": [Circular],
                      "type": "string",
                      "value": "type",
                    },
                    Object {
                      "length": 7,
                      "offset": 36,
                      "parent": [Circular],
                      "type": "string",
                      "value": "input",
                    },
                  ],
                  "colonOffset": 34,
                  "length": 15,
                  "offset": 28,
                  "parent": [Circular],
                  "type": "property",
                },
                Object {
                  "children": Array [
                    Object {
                      "length": 9,
                      "offset": 51,
                      "parent": [Circular],
                      "type": "string",
                      "value": "binding",
                    },
                    Object {
                      "length": 11,
                      "offset": 62,
                      "parent": [Circular],
                      "type": "string",
                      "value": "some.data",
                    },
                  ],
                  "colonOffset": 60,
                  "length": 22,
                  "offset": 51,
                  "parent": [Circular],
                  "type": "property",
                },
                [Circular],
              ],
              "length": 165,
              "offset": 5,
              "type": "object",
            },
            "type": "property",
          },
          "type": "object",
        },
        "type": "property",
      },
      "type": "object",
    },
    "type": "missing",
  },
]
`;

exports[`Export Test Exports Typescript Types With Filters 1`] = `
"import { Expression, Asset, Binding, AssetWrapper } from \\"@player-ui/types\\";

/**
 * This is the most generic way of gathering data. The input is bound to a data model using the 'binding' property.
 * Players can get field type information from the 'schema' definition, thus to decide the input controls for visual rendering.
*/
export interface InputAsset<AnyTextAsset extends Asset = Asset> {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"input\\";
    /** Asset container for a field label. */
    label?: AssetWrapper<AnyTextAsset>;
    /** The location in the data-model to store the data */
    binding: Binding;
    /** Optional additional data */
    metaData?: {
        /** Additional data to beacon when this input changes */
        beacon?: string | Record<string, any>;
    };
    [key: string]: unknown;
}
export interface TextAsset {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"text\\";
    /** The text to display */
    value: string;
    /** Any modifiers on the text */
    modifiers?: Array<{
        /** The modifier type */
        type: string;
        /** Modifiers can be named when used in strings */
        name?: string;
        [key: string]: unknown;
    } | {
        /** The link type denotes this as a link */
        type: \\"link\\";
        /** An optional expression to run before the link is opened */
        exp?: Expression;
        /** metaData about the link's target */
        metaData: {
            /** The location of the link to load */
            ref: string;
            /** Used to indicate an application specific resolver to use */
            'mime-type'?: string;
        };
    }>;
    [key: string]: unknown;
}
/**
 * User actions can be represented in several places.
 * Each view typically has one or more actions that allow the user to navigate away from that view.
 * In addition, several asset types can have actions that apply to that asset only.
*/
export interface ActionAsset<AnyTextAsset extends Asset = Asset> {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"action\\";
    /** The transition value of the action in the state machine */
    value?: string;
    /** A text-like asset for the action's label */
    label?: AssetWrapper<AnyTextAsset>;
    /** An optional expression to execute before transitioning */
    exp?: Expression;
    /** An optional string that describes the action for screen-readers */
    accessibility?: string;
    /** Additional optional data to assist with the action interactions on the page */
    metaData?: {
        /** Additional data to beacon */
        beacon?: string | Record<string, any>;
        /** Force transition to the next view without checking for validation */
        skipValidation?: boolean;
    };
    [key: string]: unknown;
}
export interface InfoAsset {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"info\\";
    /** The string value to show */
    title?: AssetWrapper;
    /** subtitle */
    subTitle?: AssetWrapper;
    /** Primary place for info */
    primaryInfo?: AssetWrapper;
    /** List of actions to show at the bottom of the page */
    actions?: Array<AssetWrapper>;
    [key: string]: unknown;
}
export interface CollectionAsset {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"collection\\";
    /** The string value to show */
    values: Array<AssetWrapper>;
    [key: string]: unknown;
}"
`;

exports[`Export Test Exports Typescript Types With Transforms 1`] = `
"import { Expression, Asset, Binding, AssetWrapper } from \\"@player-ui/types\\";

/**
 * This is the most generic way of gathering data. The input is bound to a data model using the 'binding' property.
 * Players can get field type information from the 'schema' definition, thus to decide the input controls for visual rendering.
*/
export interface InputAsset<AnyTextAsset extends Asset = Asset> {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"input\\";
    /** Asset container for a field label. */
    label?: AssetWrapper<AnyTextAsset>;
    /** The location in the data-model to store the data */
    binding: Binding;
    /** Optional additional data */
    metaData?: {
        /** Additional data to beacon when this input changes */
        beacon?: string | Record<string, any>;
    };
    transformed?: true;
    [key: string]: unknown;
}
export interface TextAsset {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"text\\";
    /** The text to display */
    value: string;
    /** Any modifiers on the text */
    modifiers?: Array<{
        /** The modifier type */
        type: string;
        /** Modifiers can be named when used in strings */
        name?: string;
        [key: string]: unknown;
    } | {
        /** The link type denotes this as a link */
        type: \\"link\\";
        /** An optional expression to run before the link is opened */
        exp?: Expression;
        /** metaData about the link's target */
        metaData: {
            /** The location of the link to load */
            ref: string;
            /** Used to indicate an application specific resolver to use */
            'mime-type'?: string;
        };
    }>;
    transformed?: true;
    [key: string]: unknown;
}
/**
 * User actions can be represented in several places.
 * Each view typically has one or more actions that allow the user to navigate away from that view.
 * In addition, several asset types can have actions that apply to that asset only.
*/
export interface ActionAsset<AnyTextAsset extends Asset = Asset> {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"action\\";
    /** The transition value of the action in the state machine */
    value?: string;
    /** A text-like asset for the action's label */
    label?: AssetWrapper<AnyTextAsset>;
    /** An optional expression to execute before transitioning */
    exp?: Expression;
    /** An optional string that describes the action for screen-readers */
    accessibility?: string;
    /** Additional optional data to assist with the action interactions on the page */
    metaData?: {
        /** Additional data to beacon */
        beacon?: string | Record<string, any>;
        /** Force transition to the next view without checking for validation */
        skipValidation?: boolean;
    };
    transformed?: true;
    [key: string]: unknown;
}
export interface InfoAsset {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"info\\";
    /** The string value to show */
    title?: AssetWrapper;
    /** subtitle */
    subTitle?: AssetWrapper;
    /** Primary place for info */
    primaryInfo?: AssetWrapper;
    /** List of actions to show at the bottom of the page */
    actions?: Array<AssetWrapper>;
    transformed?: true;
    [key: string]: unknown;
}
export interface CollectionAsset {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"collection\\";
    /** The string value to show */
    values: Array<AssetWrapper>;
    transformed?: true;
    [key: string]: unknown;
}"
`;

exports[`Export Test Exports Typescript types 1`] = `
"import { Expression, Asset, Binding, AssetWrapper } from \\"@player-ui/types\\";

/**
 * This is the most generic way of gathering data. The input is bound to a data model using the 'binding' property.
 * Players can get field type information from the 'schema' definition, thus to decide the input controls for visual rendering.
*/
export interface InputAsset<AnyTextAsset extends Asset = Asset> {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"input\\";
    /** Asset container for a field label. */
    label?: AssetWrapper<AnyTextAsset>;
    /** The location in the data-model to store the data */
    binding: Binding;
    /** Optional additional data */
    metaData?: {
        /** Additional data to beacon when this input changes */
        beacon?: string | Record<string, any>;
    };
    [key: string]: unknown;
}
export interface TextAsset {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"text\\";
    /** The text to display */
    value: string;
    /** Any modifiers on the text */
    modifiers?: Array<{
        /** The modifier type */
        type: string;
        /** Modifiers can be named when used in strings */
        name?: string;
        [key: string]: unknown;
    } | {
        /** The link type denotes this as a link */
        type: \\"link\\";
        /** An optional expression to run before the link is opened */
        exp?: Expression;
        /** metaData about the link's target */
        metaData: {
            /** The location of the link to load */
            ref: string;
            /** Used to indicate an application specific resolver to use */
            'mime-type'?: string;
        };
    }>;
    [key: string]: unknown;
}
/**
 * User actions can be represented in several places.
 * Each view typically has one or more actions that allow the user to navigate away from that view.
 * In addition, several asset types can have actions that apply to that asset only.
*/
export interface ActionAsset<AnyTextAsset extends Asset = Asset> {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"action\\";
    /** The transition value of the action in the state machine */
    value?: string;
    /** A text-like asset for the action's label */
    label?: AssetWrapper<AnyTextAsset>;
    /** An optional expression to execute before transitioning */
    exp?: Expression;
    /** An optional string that describes the action for screen-readers */
    accessibility?: string;
    /** Additional optional data to assist with the action interactions on the page */
    metaData?: {
        /** Additional data to beacon */
        beacon?: string | Record<string, any>;
        /** Force transition to the next view without checking for validation */
        skipValidation?: boolean;
    };
    [key: string]: unknown;
}
export interface InfoAsset {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"info\\";
    /** The string value to show */
    title?: AssetWrapper;
    /** subtitle */
    subTitle?: AssetWrapper;
    /** Primary place for info */
    primaryInfo?: AssetWrapper;
    /** List of actions to show at the bottom of the page */
    actions?: Array<AssetWrapper>;
    [key: string]: unknown;
}
export interface CollectionAsset {
    /** Each asset requires a unique id per view */
    id: string;
    /** The asset type determines the semantics of how a user interacts with a page */
    type: \\"collection\\";
    /** The string value to show */
    values: Array<AssetWrapper>;
    [key: string]: unknown;
}"
`;
